### ISSUE TITLE
Benchmark improvements and feature requests

### ISSUE THREAD

**Issue by horenmar** [catchorg/Catch2#1718]
**Description**
This is an overarching issue for Benchmarking-related feature requests. It should serve as a central place to discuss possible extensions of the current benchmarking support and figuring what the users want.

**Additional context**
Currently opened issues/requests:
* [Let user set number of collected samples and in-sample iteration](https://github.com/catchorg/Catch2/issues/1711)
* [Let user divide per-iteration timing](https://github.com/catchorg/Catch2/issues/1690) -- Here the idea is that each iteration might be working on an N element batch, and thus the reported iteration times should be divided further.
* [Let user control the measured time](https://github.com/catchorg/Catch2/pull/1701) -- This could also be solved by letting the user ask for more iterations/telling user of more iterations.

**Comment by RT222** [comment]
Good idea. I'd like to add another suggestion : It would be nice to allow the user to customize the benchmarks ouput style. I find the current formatting of the results not very easy to read, and liked the table style of the previous version better.

**Comment by jimpark** [comment]
Can we make running benchmarks optional and only enabled via cli option? The problem is that running tests now also run benchmarks which take a lot longer to run. I'd like the default behavior of CTest (CMake tests) to run without benchmarking to save build time since unit tests are run with the build.

**Comment by poconbhui** [comment]
There was some discussion on the Discord server, but I'd like to bring it up here, about the pros and cons of allowing users to set their own elapsed timings, as in #1701 .

I'll start things off, how I see them.

## Pros
* Allows user to arbitrarily alter timing
* No need to add extra abstractions as use cases arise
  * i.e. allows users to implement the motivation behind both #1690 (per-iteration timing for benchmarked loops) and #1701 (consistent number of calls to benchmark body in distributed and load-imbalanced code)
* May encourage messy / unintuitive code

## Cons
* Allows user to arbitrarily alter timing
* Requires exposing some internals of the timing (i.e. units of time measurable by the clock), potentially effecting maintainability + portability.
* Requires user to understand something of the benchmark timing and use sane timing modifications.
* Requires some extra thought to ensure the measurement phase of the benchmark works as expected when users can modify elapsed times.

I have plenty of biased comments to add, but I'll let someone else comment first!

**Comment by gian21391** [comment]
My suggestion is to give the possibility to include custom columns in the results. Something like [custom counters](https://github.com/google/benchmark#custom-counters) in Google Benchmark.

**Comment by RT222** [comment]
I have two more suggestions :

- Replace --benchmark-no-analysis with --benchmark-analysis, thus disabling analysis by default. I think it would be a better default option because, while being a very cool feature to have, benchmark analysis is slower, harder to read and probably not very useful in a majority of cases.

- Add baselines. It's a very convenient feature to have in a benchmark toolset in order to compare your code performance to a reference.

**Comment by VioletGiraffe** [comment]
I have not necessarily a request, but a question at first: how do I benchmark code with side effects where each iteration needs to repeat, but not measure, the preparation step?
Say I want to measure std::vector::clear():

```
BENCHMARK_ADVANCED("clear")(Catch::Benchmark::Chronometer meter) {
			std::vector<int> v(100000, 0);
			meter.measure([&] { v.clear(); return v.size(); });
		};
```

This is wrong because `measure` will be called multiple times, and all but the very first call are no-ops.

**Comment by Dudly01** [comment]
@horenmar what @VioletGiraffe says is something I came across too as a problem as I want to benchmark a function that modifies its input argument. I tried to solve it and this is what I could think of:
- I thought about inserting the _preparation step_ into the measurement and measuring it separately to just substract the result. This feels troublesome if the _preparation step_ consits of setting up more than one variables (or copying those from before the `meter.measure([&] { ... });` part.
- Google benchmark [has a way to overcome this](https://github.com/google/benchmark#controlling-timers). I just do not like using it instead of Catch as I feel like it is cluttering my project.
- Maybe being able to set the runs per samples could also overcome this. (At least for me, as my code is only running once per sample during measurement).

Any insight to this problem?

**Comment by toughengineer** [comment]
My understanding is that Catch uses wall clock to measure time (which appears to be `std::chrono::steady_clock`).
Using wall clock makes benchmarking results wildly noisy in ordinary situtation (i.e. without fixing CPU frequency, disabling SMT, etc., which is a pain in the butt to do and undo every time you want to benchmark something), which is OK for quick and dirty benchmarking and "eyeball statistics" evaluation (i.e. nonrigorous, unscientific evaluation) which is good enough in many cases.

Google benchmark provides CPU time which is measured by OS (e.g. `GetProcessTimes()`, `getrusage()`, see [google benchmark code](https://github.com/google/benchmark/blob/f02794a8bda1295ed7a6c99615f8b79e6188d7f8/src/timers.cc#L118)) which gives more information that, among other things, allows to better estimate the reliability of the results and diagnose measurement deviations/noise.

It would be wonderful to have CPU time estimation in Catch as well, similar to what google benchmark does, it would increase usefulness of benchmarking functionality very much, _especially_ when the environment is not properly prepared (i.e. without fixing CPU frequency, etc., I bet that's _the majority_ of real life use cases) essentially making it more immediately available.
